libyul/optimiser/UnusedFunctionParameterPruner.cpp

```diff
@@ -0,0 +1,136 @@
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * UnusedFunctionParameterPruner: Optimiser step that removes unused parameters from function
+ * definition.
+ */
+
+#include <libyul/optimiser/UnusedFunctionParameterPruner.h>
+#include <libyul/optimiser/UnusedFunctionsCommon.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDisplacer.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/YulString.h>
+#include <libyul/AsmData.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <algorithm>
+#include <optional>
+#include <variant>
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+using namespace solidity::yul::unusedFunctionsCommon;
+
+void UnusedFunctionParameterPruner::run(OptimiserStepContext& _context, Block& _ast)
+{
+	map<YulString, size_t> references = ReferencesCounter::countReferences(_ast);
+	auto used = [&](auto v) -> bool { return references.count(v.name); };
+
+	// Function name and a pair of boolean masks, the first corresponds to parameters and the second
+	// corresponds to returnVariables.
+	//
+	// For the first vector in the pair, a value `false` at index `i` indicates that the function
+	// argument at index `i` in `FunctionDefinition::parameters` is unused inside the function body.
+	//
+	// Similarly for the second vector in the pair, a value `false` at index `i` indicates that the
+	// return parameter at index `i` in `FunctionDefinition::returnVariables` is unused inside
+	// function body.
+	map<YulString, pair<vector<bool>, vector<bool>>> usedParametersAndReturnVariables;
+
+	// Step 1 of UnusedFunctionParameterPruner: Find functions whose parameters (both arguments and
+	// return-parameters) are not used in its body.
+	for (auto const& statement: _ast.statements)
+		if (holds_alternative<FunctionDefinition>(statement))
+		{
+			FunctionDefinition const& f = std::get<FunctionDefinition>(statement);
+
+			if (tooSimpleToBePruned(f))
+				continue;
+
+			vector<bool> parameters = applyMap(f.parameters, used);
+			vector<bool> returnVariables = applyMap(f.returnVariables, used);
+
+			pair<vector<bool>, vector<bool>> parametersAndReturnVariables{
+				vector<bool>(f.parameters.size(), true),
+				vector<bool>(f.returnVariables.size(), true)
+			};
+
+			bool hasUnused = false;
+			if (anyFalse(parameters))
+			{
+				parametersAndReturnVariables.first = move(parameters);
+				hasUnused = true;
+			}
+			if (anyFalse(returnVariables))
+			{
+				parametersAndReturnVariables.second = move(returnVariables);
+				hasUnused = true;
+			}
+
+			if (hasUnused)
+				usedParametersAndReturnVariables[f.name] = move(parametersAndReturnVariables);
```

Comment:
```suggestion
			if (tooSimpleToBePruned(f) || boost::algorithm::all_of(f.parameters + f.returnVariables, used))
				continue;

			usedParametersAndReturnVariables[f.name] = {
				applyMap(f.parameters, used),
				applyMap(f.returnVariables, used)
			};
```
This evaluates `used` twice, so you could also introduce local variables, but I don't think we should be shy of using a concatenation of vectors of bools...